### PR TITLE
Feature/On-prem scanners: Run parallel scans with pull consumers.

### PR DESCRIPTION
**Pull request by deadly-panda**

### CODE REVIEW

File: src/ostorlab/scanner/scan_handler.py [hunk 1]
```diff
@@ -40,35 +39,83 @@ async def _subscribe(
...
+        The message is acknowledged after the scan is created, and the scan handler can't fetch new messages,
+        until the current scan is no longer running.
+
+        Args:
+            bus_handler: instance for performing BUS operations.
+            config: The scanner configuration; holds credentials for the registry & streaming server.
+        """
+        is_scan_running = False
+        while True:
+            if is_scan_running is True:
```
Comment (3asm) [inline src/ostorlab/scanner/scan_handler.py:89]: This code will never exist.

File: src/ostorlab/cli/scanner/scanner.py [hunk 2]
```diff
@@ -54,32 +57,52 @@ def scanner(
     state_reporter = scanner_state_reporter.ScannerStateReporter(
         scanner_id=scanner_id, hostname=socket.gethostname(), ip=ip.get_ip()
     )
+    nb_parallel_scans = int(n)
```
Comment (amine3) [inline src/ostorlab/cli/scanner/scanner.py:60]: we need to check that nb_parallel_scans > 0

**Review by amine3 — CHANGES_REQUESTED** [review]

File: src/ostorlab/cli/scanner/scanner.py (hunk 2 above)
Comment (deadly-panda) [reply src/ostorlab/cli/scanner/scanner.py:60]: It is already the case: `type=click.IntRange(1, None)` will make the sub-command fail if it is `< 0`.

File: src/ostorlab/scanner/handler.py [hunk 3]
```diff
@@ -198,28 +193,36 @@ async def subscribe(
...
+    async def process_message(
+        self,
+    ):
+        msg = None
+        request = None
+
+        if self._pull_subscription is not None:
+            try:
+                msg = await self._pull_subscription.fetch()
+                msg = msg[0]
```
Comment (3asm) [inline src/ostorlab/scanner/handler.py:216]: Why msg[0]?

File: src/ostorlab/scanner/scan_handler.py [hunk 4]
```diff
@@ -40,35 +39,83 @@ async def _subscribe(
...
+                    # No available message to fetch.
+                    await asyncio.sleep(WAIT_CHECK_MESSAGES.seconds)
+
+    async def _handle_msg(
+        self,
+        bus_handler: scanner_handler.BusHandler,
+        config: scanner_conf.ScannerConfig,
+    ) -> str:
+        """Fetch, parse a single message and trigger the corresponding scan."""
+        msg, request = await bus_handler.process_message()
```
[the remark this answers is not in the document]
Comment (3asm) [reply src/ostorlab/scanner/scan_handler.py:104]: Ping.

File: src/ostorlab/scanner/scan_handler.py [hunk 5]
```diff
@@ -40,35 +39,82 @@ async def _subscribe(
...
+        The message is acknowledged after the scan is created, and the scan handler can't fetch new messages,
+        until the current scan is no longer running.
+
+        Args:
+            bus_handler: instance for performing BUS operations.
+            config: The scanner configuration; holds credentials for the registry & streaming server.
+        """
+        scan_id = None
+        while True:
+            if _is_scan_running(self._docker_client, scan_id=scan_id):
```
Comment (3asm) [inline src/ostorlab/scanner/scan_handler.py:89]: Explicit check.

File: src/ostorlab/scanner/scan_handler.py [hunk 6]
```diff
@@ -40,35 +39,82 @@ async def _subscribe(
...
                 config=config,
                 subject=bus_conf.subject,
                 queue=bus_conf.queue,
-                cb=functools.partial(
-                    callbacks.cb_start_scan,
-                    state_reporter=self._state_reporter,
-                    registry_conf=config.registry_conf,
-                ),
-                start_at="last_received",
+                start_at="first",
```
Comment (3asm) [inline src/ostorlab/scanner/scan_handler.py:64]: Why has this changed from last_received to first?

File: src/ostorlab/scanner/handler.py [hunk 7]
```diff
@@ -70,7 +70,7 @@ async def connect(self, connect_timeout=DEFAULT_CONNECT_TIMEOUT.seconds):
             name=self._name,
             tls=self._tls_context,
             connect_timeout=connect_timeout,
-            error_cb=None,
+            error_cb=None, #self._error_cb,
```
Comment (3asm) [inline src/ostorlab/scanner/handler.py]: Why remove the error_cb?

**Review by 3asm — CHANGES_REQUESTED** [review]

**Pull request merged**